### PR TITLE
Move Cursor field to SupportPower.

### PR DESCRIPTION
**Pull request by phrohdoh**

### CODE REVIEW

**Comment by pchote** [comment]
Looks like a reasonable change, but not tested. :+0.5:

**Comment by Mailaender** [comment]
You forgot to update the d2k/ra and ts rules. They also have nuke icons.

**Comment by Phrohdoh** [comment]
Good catch, fixed.

**Comment by Mailaender** [comment]
:+1:

**Comment by pchote** [comment]
That will also be a problem for mod maps and mods. Please add an upgrade rule.

**Comment by Phrohdoh** [comment]
I wrote this but it isn't working:

```diff
﻿﻿diff --git a/OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs b/OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs
index a2d6711..6fd3e1f 100644
--- a/OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs
+++ b/OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs
@@ -1311,6 +1311,33 @@ internal static void UpgradeWeaponRules(int engineVersion, ref List<MiniYamlNode
 					}
 				}
 
+				if (engineVersion < 20150501)
+				{
+					var isNukePower = node.Key == "NukePower";
+					var isIonCannonPower = node.Key == "IonCannonPower";
+
+					if (isNukePower || isIonCannonPower)
+					{
+						if (node.Value.Nodes.Any(n => n.Key == "Cursor"))
+							Console.WriteLine("Found cursor on {0}", node.Location);
+
+						foreach (var n in node.Value.Nodes)
+						{
+							if (n.Key == "Cursor")
+								break;
+
+							var cursor = "<bogus>";
+
+							if (isNukePower)
+								cursor = "nuke";
+							else if (isIonCannonPower)
+								cursor = "ioncannon";
+
+							node.Value.Nodes.Add(new MiniYamlNode("Cursor", cursor));
+						}
+					}
+				}
+
 				UpgradeWeaponRules(engineVersion, ref node.Value.Nodes, node, depth + 1);
 			}
 		}
```

**Comment by pchote** [comment]
That `break` will trash the upgrader state.  I suggest negating it and moving it into the conditional.

**Comment by pchote** [comment]
You already know that one of `isNukePower` and `isIonCannonPower` is true, so you can use `var cursor = isIonCannonPower ? "ioncannon" : "nuke";`.

**Comment by Phrohdoh** [comment]
Updated.

**Comment by penev92** [comment]
Since those two traits no longer have their own Cursors, things change in their `OrderGenerator` methods.
The `OrderGenerator` in `IonCannonPower` no longer needs to cast `var info = Info as IonCannonPowerInfo;`.
The `OrderGenerator` in `NukePower` becomes identical to the one in `SupportPower`, so you can just remove it.

Apart from `// Clear selection if using Left-Click Orders`, the method in `IonCannonPower` is also identical to the one in the base class. I find the fact that the IonCannon cares about Left-click orders, but Nuke doesn't quite strange. I think this should be looked into, and possibly other support powers should be made to care about left-click orders?

**Comment by penev92** [comment]
Also needs a rebase now.

**Comment by Phrohdoh** [comment]
Updated and rebased.

File: OpenRA.Mods.Cnc/Traits/SupportPowers/IonCannonPower.cs [hunk 1]
```diff
@@ -50,7 +55,6 @@ public override IOrderGenerator OrderGenerator(string order, SupportPowerManager
 				manager.Self.World.Selection.Clear();
 
 			Sound.PlayToPlayer(manager.Self.Owner, Info.SelectTargetSound);
-			var info = Info as IonCannonPowerInfo;
 			return new SelectGenericPowerTarget(order, manager, info.Cursor, MouseButton.Left);
 		}
```
Comment (obrakmann) [inline OpenRA.Mods.Cnc/Traits/SupportPowers/IonCannonPower.cs]: You can indeed now remove this whole method. The selection is cleared again in the constructor of `SelectGenericPowerTarget`.

**Comment by obrakmann** [comment]
:+1: after you've fixed travis and removed the `OrderGenerator` from the `IonCannonPower`

**Comment by Phrohdoh** [comment]
Fixed.

**Comment by obrakmann** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/3a1c52437fb5df4c79baca054a95a845fc81f378%5E...3a1c52437fb5df4c79baca054a95a845fc81f378)

**Pull request merged**